qutip/solver/integrator.py

```diff
@@ -0,0 +1,185 @@
+""" `Integrator`: ODE solver wrapper to use in qutip's Solver """
+
+
+__all__ = ['Integrator', 'IntegratorException', 'add_integrator',
+           'sesolve_integrators', 'mesolve_integrators', 'mcsolve_integrators']
+
+
+from itertools import product
+from qutip import QobjEvo, qeye, basis
+from functools import partial
+
+
+class IntegratorException(Exception):
+    pass
+
+
+class Integrator:
+    """
+    A wrapper around ODE solvers.
+    It ensures a common interface for Solver usage.
+    It takes and return states as :class:`qutip.core.data.Data`, it may return
+    a different data-type than the input type.
+
+    Parameters
+    ----------
+    sytem: qutip.QobjEvo
+        Quantum system in which states evolve.
+
+    options: qutip.SolverOptions
+        Options for the solver.
+    """
+    # Used options in qutip.SolverOdeOptions
+    used_options = []
+    # Can evolve time dependent system
+    support_time_dependant = None
+    # Use the QobjEvo's matmul_data method as the driving function
+    use_QobjEvo_matmul = None
```

Comment:
- Since Solver also used option, I thought it would be clearer if self.option was always the same thing, but no they don't need all of options. If we define `self.options` to be a `dict` of only the needed options, we remove a lot of duplication, good idea.

- `support_time_dependent` is used in test, only the [`IntegratorDiag`](https://github.com/qutip/qutip/pull/1644/files#diff-96d5a6258324bcf45eb2b4cdb89e8b15dafa8899f26152046d39c4b60b82f8c2R70), does not support it for now and it check that the system is constant.

- When `use_QobjEvo_matmul` is False, feedback, adaptative hilbert space and other modification to `matmul` will fail. I am thinking that it will also be enough to tell if we can expect auto differentiation to work with the `Integrator`. If we have an Integrator that does not use the `QobjEvo.matmul_data` it will probably be unable to detect it and produce wrong result. It could be part of `add_integrator` and we keep a set of operator that support feedback and one that don't, but adding a flag here is simpler. I expect a strong correlation between `support_time_dependent` and `use_QobjEvo_matmul` but not always. Looking at `fsesolve`, it could be an `Integrator` for `sesolve` that support time-dependant (periodic) system but not feedback.